wgpu-native/src/lib.rs

```diff
@@ -168,46 +168,157 @@ pub struct ByteArray {
     pub length: usize,
 }
 
-pub type InstanceId = hub::Id;
+macro_rules! transparent {
+    ($i:item) => (
+        #[repr(transparent)]
+        #[derive(Clone, Copy, Debug, Hash, PartialEq)]
+        #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+        $i
+    )
+}
+
+pub trait ToId {
+    fn id(&self) -> hub::Id;
+}
+
+macro_rules! to_id {
+    ($i:ident) => (
+        impl ToId for $i {
+            fn id(&self) -> hub::Id {
+                self.0
+            }
+        }
+    )
+}
+
+macro_rules! from_id {
+    ($i:ident) => (
+        impl From<hub::Id> for $i {
+            fn from(id:hub::Id) -> $i {
+                $i(id)
+            }
+        }
+    )
+}
+
+use hub::{Index, Epoch, NewId, Id};
+macro_rules! new_id {
+    ($i:ident) => (
+        impl NewId for $i {
+            fn new(index: Index, epoch: Epoch) -> Self {
+                let id = Id::new(index, epoch);
+                $i(id)
+            }
+
+            fn index(&self) -> Index {
```

Comment:
I think we can omit it and just make `trait NewId: TypedId`

Reply:
I am not sure to understand, you propose to integrate the `NewId` into the `TypedId` trait?

Reply:
Nevermind, I think your approach is fine